src/contract/method.rs

```diff
@@ -16,9 +16,64 @@ use std::pin::Pin;
 use std::task::{Context, Poll};
 use web3::api::Web3;
 use web3::contract::tokens::Detokenize;
+use web3::contract::Error as Web3ContractError;
 use web3::types::{Address, BlockNumber, Bytes, CallRequest, U256};
 use web3::Transport;
 
+/// A void type to represent methods with empty return types.
+///
+/// This is used to work around the fact that `(): !Detokenize`.
+pub struct Void(());
+
+/// Represents a type can detokenize a result.
+pub trait Detokenizable {
+    /// The output that this type detokenizes into.
+    type Output;
+
+    /// Returns true if this is an empty type.
+    fn is_empty() -> bool;
+
+    /// Create an instance of `Output` by decoding tokens.
+    fn from_tokens(tokens: Vec<Token>) -> Result<Self::Output, ExecutionError>;
+}
+
+impl Detokenizable for Void {
+    type Output = ();
+
+    fn is_empty() -> bool {
+        true
+    }
+
+    fn from_tokens(tokens: Vec<Token>) -> Result<Self::Output, ExecutionError> {
+        if !tokens.is_empty() {
+            return Err(Web3ContractError::InvalidOutputType(format!(
+                "Expected no elements, got tokens: {:?}",
+                tokens
+            ))
+            .into());
+        }
+
+        Ok(())
+    }
+}
+
+impl<T: Detokenize> Detokenizable for T {
+    type Output = Self;
+
+    fn is_empty() -> bool {
```

Comment:
I find this name a little bit misleading as e.g. a string or vector could also be empty. Is this really just to see if this type can be used to symbolize an error? Maybe we could just call it is Void (or have a more dynamic way of downcasting the concrete type into void)?

Which makes me wonder what happens in geth when we revert without a message on a void function. Is it distinguishable from a success?

Reply:
> Maybe we could just call it is Void

I like `is_void`.

> Which makes me wonder what happens in geth when we revert without a message on a void function. Is it distinguishable from a success?

No, and that is sadly why we need it :(

Reply:
I renamed the method to `is_void` and added a note in the comments about how it cannot be distringuised from an `INVALID` opcode or `REVERT` without data.